### DISCUSSION TITLE
How to write my own kernel about surface tension term

### DISCUSSION THREAD

**Discussion by Wolke926** [idaholab/moose#24976] [Q&A General] [answered]
Hi all,
I am a new MOOSE user and trying to compile my own kernel, I have already gone through the examples on the MOOSE tutorial homepage but am still confused about the source code.

The exact problem I want to solve is similar to @liang-tech so I am trying to understand the code here:
[https://github.com/idaholab/moose/discussions/23382](url)
The key point in his reference paper doi:10.1103/PhysRevE.106.015111 is the surface tension part in Navier-stokes equation combined with Cahn Hilliard equation, fr = psi * grad phi (psi is the chemical potential and phi denotes the phase)

Some questions from my side are listed:
1. when should we calculate qp Jacobian? Why in the following example there are only "compute QP residual"?
[https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial01_app_development/step03_moose_object.html](url)
3. what is off-diagonal Jacobian and when do we need to calculate it? Based on the link above, there is no off-diagonal Jacobian. Is there any introduction about what these variables (such as jvar, jac) mean?
 
4. how do we decide which kind of other base objects to heritage? Because in the tutorial, it shows " ADDiffusion : public ADKernelGrad" and I can only find these examples. 
![Screen Shot 2023-07-17 at 2 12 55 PM](https://github.com/idaholab/moose/assets/120539827/0d66b37e-8645-4325-b864-f125dc8abaae)

thanks for your help!

**Comment by lindsayad** [comment, marked as answer]
> * when should we calculate qp Jacobian? Why in the following example there are only "compute QP residual"?
>   [https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial01_app_development/step03_moose_object.html](url)

In that example, the Jacobian is computed automatically using [automatic differentiation (AD)](https://mooseframework.inl.gov/automatic_differentiation/). If you use AD then you do not need to hand-code a Jacobian.

> * what is off-diagonal Jacobian and when do we need to calculate it? Based on the link above, there is no off-diagonal Jacobian. Is there any introduction about what these variables (such as jvar, jac) mean?

With AD no Jacobian routines are necessary. I would recommend starting with AD and then once you are modeling your physics successfully you can worry about optimizing code speed and hand-coding your Jacobians.

> * how do we decide which kind of other base objects to heritage? Because in the tutorial, it shows " ADDiffusion : public ADKernelGrad" and I can only find these examples.

If the weak form of your term only involves the test function and not its gradient, then you can use `ADKernelValue`. If you only use the test function gradient and not its value, then you use `ADKernelGrad`.  If it uses both gradient and value, then you use `ADKernel`.
